lib/internal/Magento/Framework/Module/Dir/Resolver.php

```diff
@@ -0,0 +1,49 @@
+<?php
+/**
+ * Copyright © 2015 Magento. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Framework\Module\Dir;
+
+/**
+ * Provides ability to statically register modules which do not reside in the modules directory
+ *
+ * @author Josh Di Fabio <joshdifabio@gmail.com>
+ */
+final class Resolver implements ResolverInterface
+{
+    /**
+     * Paths to modules
+     *
+     * @var string[]
+     */
+    private static $modulePaths = [];
+
+    /**
+     * Sets the location of a module. Necessary for modules which do not reside in modules directory
+     *
+     * @param string $moduleName Fully-qualified module name
+     * @param string $path Absolute file path to the module
+     */
+    public static function setModulePath($moduleName, $path)

```

Comment:
@antonkril thanks for reviewing this.

> Static calls are hard-coded dependencies, they can not be substituted or decorated. So please avoid them.

I agree with you, and I avoid them wherever possible and appropriate. Sometimes, however, it is necessary to use them. Refer to one of the static methods in Magento 2 for an example.

In this case, the only way to achieve what I've proposed in this PR is with a static method call. Please take a look at the section entitiled *What would a module look like with this approach?* in the OP to see why.

Furthermore, I consider a module to be a static entity in the same way that a class is a static entity. While a class has a single location on disk, so does a module. Consider that autoloaders are registered in PHP statically (via `spl_autoload_register`). Some functionality *is* static. This is essentially autoloading for modules; the reason we need to implement it ourselves here is that packages/modules aren't first-class citizens in PHP as classes are.

Also note that nothing in Magento depends on that static method call; it doesn't affect testability.

> Other problem is that with such approach there is no strict rule when all module paths are available, because you don't know if all modules called setModulePaths, and module dev doesn't know when it should call setModulePath.

I'm not sure I understand the argument. If a module doesn't reside in `app/code` and doesn't enable itself via the module resolver then yes, it won't be enabled. Similarly, it won't be enabled if it doesn't implement a `module.xml`. I don't see what's strange about that.